feat: Add total views for each forum post #23

thetalesofj opened the pull request:
# Description

<!-- Please include a detailed summary of the changes made. Also please link the issue that this PR is fixing. -->

Fixes #23 

<!-- Please place an x in the [ ] to check off the type of change for this PR -->

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Chore items (this includes basic clean up of files, or updates to packages)
- [ ] Updates to documentation

<!-- Please make sure to go through the entire checklist before requesting a review. Place an x in the [ ] to check off all of the items completed -->

## Checklist

- [ x] I have a descriptive title for my PR
- [ x] I have linked the issue to this [PR](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x ] I have run the project locally and reviewed the code changes
- [x ] My changes generate no new warnings

Dorfieeee commented on helpers.js:
```diff
@@ -36,3 +36,16 @@ export const formatDateDiff = (recent, old) => {
+export function formatLargeNumber(n) {
+  let s = "";
+  if (n < 1000) {
+    s = n;
```
The function must return `string`.
When `n` is < than 1000, the function returns an `integer`.

Dorfieeee commented on helpers.js:
```diff
@@ -36,3 +36,16 @@ export const formatDateDiff = (recent, old) => {
+  if (n < 1000) {
+    s = n;
+  } else if (n >= 1000 && n < 1000000) {
+    s = (Math.round(n)/1000).toFixed(1) + "k"
```
This does not work when `n = 999,999`. The expected result is `999k` but actual result is `1000.0k`
The same problem with `n = 499,950` => `499.9k`, but `n = 499,951` => `500.0k`.

Dorfieeee commented on helpers.js:
```diff
@@ -36,3 +36,16 @@ export const formatDateDiff = (recent, old) => {
+  } else if (n >= 1000 && n < 1000000) {
+    s = (Math.round(n)/1000).toFixed(1) + "k"
+  } else if (n >= 1000000) {
+    s = (Math.round(n)/1000000).toFixed(1) + "M"
```
Same problem as above, only with other numbers > 1,000,000.

thetalesofj replied:
Using `Math.trunc` instead of `Math.round` e.g. `(Math.trunc(n)/1000) + "k"` resolves the issue of 500.0k to 500k.  However, I'm seeing how I can alter this to make 499999 = 499.9k instead of 499.999k

Dorfieeee replied:
You may try to convert the result to string first, split it by '.' character and add some if else you statement whether or not you want to display the decimal place

Dorfieeee replied:
I think you're limiting yourself with the use of built-in methods and trying too much to integrate them, that when they don't provide the exact solution you need, you get burnt and don't know the way out.. and I understand.

What if you tried using simple techniques and array operations instead which I believe you can as you worked through some challenges on Codewars.

I'll give you a hint

```suggestion
    let temp = String(n/1000000).split(".");
    // add some conditional operations bellow
    // try it without accessing indices out of the array's range
    s += "M";
```

Dorfieeee reviewed (changes requested):
It would be also nice if result like `s = 500.0k` did not display the decimal part so it would be only `s = 500k`

thetalesofj commented:
If what I've committed isn't readable this could do the trick:

`function formatFig(n) { 
   let s = "";
   if (n < 1000) { 
    let temp = String(n/1).split(".");
    s+= temp 

   } else if (n >= 1000 && n < 1000000) { 
    let temp = String(n/10000).split("");
    let b =  temp[2] + temp[1] + temp[3]
    s+= b + "k"

   } else if (n >= 1000000 && n <= 9999999) { 
    let temp = String(n/1000000).split("");
    let b = temp[0] + temp[1] + temp[2]
    s+= b + "M"

   } else if (n >= 10000000 && n < 99999999) { 
    let temp = String(n/1000000).split("");
    let b = temp[0] + temp[1] + temp[2] + temp[3]
    s+= b + "M"

   } else if (n >= 100000000 && n < 999999999) { 
    let temp = String(n/1000000).split("");
    let b = temp[0] + temp[1] + temp[2] + temp[3] + temp[4]
    s+= b + "M"

   }
return s
}`

thetalesofj commented:
If what I've committed isn't readable this could do the trick:

```
function formatFig(n) {
let s = "";
if (n < 1000) {
let temp = String(n/1).split(".");
s+= temp

} else if (n >= 1000 && n < 1000000) {
let temp = String(n/10000).split("");
let b = temp[2] + temp[1] + temp[3]
s+= b + "k"

} else if (n >= 1000000 && n <= 9999999) {
let temp = String(n/1000000).split("");
let b = temp[0] + temp[1] + temp[2]
s+= b + "M"

} else if (n >= 10000000 && n < 99999999) {
let temp = String(n/1000000).split("");
let b = temp[0] + temp[1] + temp[2] + temp[3]
s+= b + "M"

} else if (n >= 100000000 && n < 999999999) {
let temp = String(n/1000000).split("");
let b = temp[0] + temp[1] + temp[2] + temp[3] + temp[4]
s+= b + "M"

}
return s
}
```

thetalesofj commented:
> If what I've committed isn't readable this could do the trick:
> 
> ```
> function formatFig(n) {
> let s = "";
> if (n < 1000) {
> let temp = String(n/1).split(".");
> s+= temp
> 
> } else if (n >= 1000 && n < 1000000) {
> let temp = String(n/10000).split("");
> let b = temp[2] + temp[1] + temp[3]
> s+= b + "k"
> 
> } else if (n >= 1000000 && n <= 9999999) {
> let temp = String(n/1000000).split("");
> let b = temp[0] + temp[1] + temp[2]
> s+= b + "M"
> 
> } else if (n >= 10000000 && n < 99999999) {
> let temp = String(n/1000000).split("");
> let b = temp[0] + temp[1] + temp[2] + temp[3]
> s+= b + "M"
> 
> } else if (n >= 100000000 && n < 999999999) {
> let temp = String(n/1000000).split("");
> let b = temp[0] + temp[1] + temp[2] + temp[3] + temp[4]
> s+= b + "M"
> 
> }
> return s
> }
> ```

Actually, Nevermind! This route starts to become more complex the more loops I need to introduce. Let me know if the commit is okay or not

Dorfieeee commented:
@thetalesofj 
This is my solution for comparison.
```
function formatLargeNumber(n) {
    let postfix = "", s = "";
    if (n > 999 && n < 1000000) {
        n = n / 1000
        postfix = "k";
    } else if (n >= 1000000) {
        n = n / 1000000
        postfix = "M";
    }
    let temp = String(n).split(".");
    s += temp[0];
    if (temp.length > 1 && temp[1][0] !== "0") {
        s += "." + temp[1][0];
    }
    s += postfix;
    return s;
}
```

thetalesofj commented:
> @thetalesofj This is my solution for comparison.
> 
> ```
> function formatLargeNumber(n) {
>     let postfix = "", s = "";
>     if (n > 999 && n < 1000000) {
>         n = n / 1000
>         postfix = "k";
>     } else if (n >= 1000000) {
>         n = n / 1000000
>         postfix = "M";
>     }
>     let temp = String(n).split(".");
>     s += temp[0];
>     if (temp.length > 1 && temp[1][0] !== "0") {
>         s += "." + temp[1][0];
>     }
>     s += postfix;
>     return s;
> }
> ```

Definitely more readable than mine. What do you think? Keep the commit the same or alter it?

Dorfieeee commented:
> Keep the commit the same or alter it?

I'll leave that up to you.. But if you can, it would be better if your function did only 1 decimal place by default without passing the extra variable.. I don't think it's necessary.. there are also cases where your function behaves strangely with very large numbers or when the parameter determining decimal place is 8 or more.

Benchmark:
![image](https://user-images.githubusercontent.com/65303812/203874837-471d0f52-dab2-44ff-b0a3-68330a29b1b0.png)

thetalesofj commented:
> > Keep the commit the same or alter it?
> 
> I'll leave that up to you.. But if you can, it would be better if your function did only 1 decimal place by default without passing the extra variable.. I don't think it's necessary.. there are also cases where your function behaves strangely with very large numbers or when the parameter determining decimal place is 8 or more.
> 
> Benchmark: ![image](https://user-images.githubusercontent.com/65303812/203874837-471d0f52-dab2-44ff-b0a3-68330a29b1b0.png)

Which software/site did you use to test the speed?
